### ISSUE TITLE
Problems with nested grid elements when migrating from fluidcontent to flux

### ISSUE THREAD

**Issue by fnagel** [FluidTYPO3/flux#1645]
I'm having a hard time migrating a TYPO3 8.7.20 instance from EXT:fluidcontent (5.2.0 with flux v8.2.1) to EXT:flux (9.1.0). 

## Migration path

Here's what I do:

* Update my template (see below for new and old version)
* Change `Ctype` using SQL: `UPDATE tt_content SET CType = 'vendor_row' WHERE CType = 'fluidcontent_content' AND tx_fed_fcefile = 'Ns.Vendor:Row.html'`
* Run the EM migration wizard in EXT:flux

## Functionality 

What happens:

* Non nested grid rows (existing and newly created ones) work as expexted
* Nested grid rows with other content elements throw error messages in backend but existing ones seem to work in frontend

## Screenshots
![2018-11-01 21_59_34-vm typo3 cms 8 7 20](https://user-images.githubusercontent.com/116824/47881690-26bfe080-de27-11e8-9234-8f91de8b8824.png)
![2018-11-01 22_05_51-vm typo3 cms 8 7 20](https://user-images.githubusercontent.com/116824/47881691-27587700-de27-11e8-82eb-874439b745cf.png)
![2018-11-01 22_09_45-vm typo3 cms 8 7 20](https://user-images.githubusercontent.com/116824/47881692-27587700-de27-11e8-95e4-f8e55875a1d4.png)

## Error messages

Error message: 
``
Argument 1 passed to FluidTYPO3\Flux\Form\Container\Grid::buildExtendedBackendLayoutArray() must be of the type integer, array given, called in /var/www/typo3conf/ext/flux/Classes/Integration/Overrides/BackendLayoutView.php on line 76 
``

Sometimes this error message pops up (example):
``
An exception occurred while executing 'SELECT * FROM `tt_content` WHERE (`tt_content`.`pid` = 649) AND (`colPos` IN (0)) AND (`sys_language_uid` IN (0, -1) AND colPos IN () ) AND ((`tt_content`.`deleted` = 0) AND ((`tt_content`.`t3ver_wsid` = 0) OR (`tt_content`.`t3ver_state` <= 0))) ORDER BY `sorting` ASC LIMIT 100': You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ') ) AND ((`tt_content`.`deleted` = 0) AND ((`tt_content`.`t3ver_wsid` = 0) OR (`' at line 1 
``

## Templates (simplified)

Old one (using EXT:fluidcontent):

	
	<f:section name="Configuration">
		<v:variable.set name="columnItems" value="{
			null: 'Default (equal width)',
			'12': 'Full width',
			'10': '5/6 width',
			'9': '3/4 width',
			'8': '2/3 width',
			'6': '1/2 width',
			'4': '1/3 width',
			'3': '1/4 width',
			'2': '1/6 width'
		}" />

		<flux:form id="row" label="Grid Row" description="Add content in columns.">
			<flux:form.option.group value="common" />
			<flux:form.option.icon>EXT:project/Resources/Public/Icons/grid.svg</flux:form.option.icon>

			<flux:form.section name="columns">
				<flux:form.object name="column" label="Column">
					<flux:field.select label="Large devices" name="class-large" items="{columnItems}" />
					<flux:field.select label="Medium devices" name="class-medium" items="{columnItems}" />
					<flux:field.select label="Small devices" name="class" items="{columnItems}" />
					<flux:field.checkbox name="center" label="Center column" />
				</flux:form.object>
			</flux:form.section>
			<flux:grid>
				<flux:grid.row>
					<v:condition.type.isArray value="{columns}">
						<f:for each="{columns}" as="sectionObject" iteration="iteration">
							<flux:grid.column name="column{iteration.cycle}" label="auto" />
						</f:for>
					</v:condition.type.isArray>
				</flux:grid.row>
			</flux:grid>
		</flux:form>
	</f:section>

New one (EXT:flux only, added `colPos` attribute to `flux:grid.column`):

	<f:section name="Configuration">
		<v:variable.set name="columnItems" value="{
			null: 'Default (equal width)',
			'12': 'Full width',
			'10': '5/6 width',
			'9': '3/4 width',
			'8': '2/3 width',
			'6': '1/2 width',
			'4': '1/3 width',
			'3': '1/4 width',
			'2': '1/6 width'
		}" />

		<flux:form id="row" label="Grid Row" description="Add content in columns.">
			<flux:form.option.group value="common" />
			<flux:form.option.icon>EXT:project/Resources/Public/Icons/grid.svg</flux:form.option.icon>

			<flux:form.section name="columns">
				<flux:form.object name="column" label="Column">
					<flux:field.select label="Large devices" name="class-large" items="{columnItems}" />
					<flux:field.select label="Medium devices" name="class-medium" items="{columnItems}" />
					<flux:field.select label="Small devices" name="class" items="{columnItems}" />
					<flux:field.checkbox name="center" label="Center column" />
				</flux:form.object>
			</flux:form.section>
			<flux:grid>
				<flux:grid.row>
					<v:condition.type.isArray value="{columns}">
						<f:for each="{columns}" as="sectionObject" iteration="iteration">
							<flux:grid.column name="column{iteration.cycle}" colPos="{iteration.cycle}" label="auto" />
						</f:for>
					</v:condition.type.isArray>
				</flux:grid.row>
			</flux:grid>
		</flux:form>
	</f:section>

Anybody able to help me out here?

**Comment by NamelessCoder** [comment]
Thanks for the detailed information, Felix!

All your migration steps are correct and the templates also look correct. The container content types also seem to resolve correctly (I'm basing that on the columns working and having labels, which means the virtual colPos values are working).

The errors however look like unhandled bad data types. I think one of them may be solved with https://github.com/FluidTYPO3/flux/commit/36e89e7b8308e46b55b1281ae95c1aaf91872025 (the SQL error) whereas the other seems to happen when some part of the record is a folded-out array (passed from FormEngine) instead of the usual integer, string or other. And I think in this case it has to do with the `colPos` field which is sometimes an array (because it is `select` type) containing a single selected integer value.

It would help if you can repeat the test using `dev-development`. A 9.2 release of Flux will follow shortly but if only one of your two issues are solved, I'd like to add a solution for the second one before that new release.

**Comment by fnagel** [comment]
I have to thank you for the quick answer! 

I've just tried `dev-development` with the same instance and data as before:

When running the wizard in EM I now get a `Gateway Timeout The gateway did not receive a timely response from the upstream server or application.`. Is there a way to run those EM wizard using the cli? I've tried increasing the PHP `max_execution_time` value to 3600 but that did not help (and is not possible on my PROD server). Any idea why this keeps happening? Was not the case with latest TER release.

Anyway, I've tested on new pages and still run into the `Argument 1 passed to FluidTYPO3\Flux\Form\Container\Grid::buildExtendedBackendLayoutArray() must be of the type integer` error. The SQL error did not pop up but that one was a little hard to reproduce before too.

Any more info / testing you need in order to come up with a fix?

**Comment by NamelessCoder** [comment]
> Is there a way to run those EM wizard using the cli?

No to my knowledge, no. Perhaps the TYPO3 console allows it, but it's quite an old implementation that seems to just be called as a user-function reference in an expected class in an expected file...

About the timeout issue, I suspect an infinite loop during processing of child elements, which could in turn be caused by inconsistent DB content combined with Flux not keeping track of which records have been migrated. Thought it would of course help to keep track and avoid infinite recursion - that would still not help you solve it, only point you to the cause (better than infinite loop!).

The other error can also be avoided by proper type checks. I'm not afraid to admit there are some cases left with unhandled types, mostly when the core neglects to declare possible value types for values passed inside arrays (e.g. UID might be a string, might be an int, might be null).

If you have the time to help with this one, it would be great with an infinite recursion prevention in the EM upgrade script combined with a check (guard-clause early-return?) to avoid the PHP error. I'm not sure that's the entire solution but both cases seem like an exception would be preferable.

Sorry this isn't super helpful. By the way: if you ever run into PHP errors encountered in Flux code it is pretty an auto-merge if you provide a PR that either throws an exception or returns early without message, if that is appropriate, to avoid the PHP error.

**Comment by fnagel** [comment]
Seems there is no way to run the EM wizards using cli. Seems hard to accomplish, see https://github.com/TYPO3-Console/TYPO3-Console/issues/662

I will try to come up with a solution for the migration wizard. Might take a few days as there some other stuff I need to work on at the moment. I'm not sure if I'm able to look at this value type thing as I'm not fully aware how flux handles stuff internally but that infinite loop thing sounds like something a flux noob should be able to handle.

I will keep you up to date!

**Comment by fnagel** [comment]
> About the timeout issue, I suspect an infinite loop during processing of child elements, which could in turn be caused by inconsistent DB content combined with Flux not keeping track of which records have been migrated (EDIT: correction, Flux does keep track - in DB - but not in the script that does the migration to prevent selecting the same record twice. This might be a question of a missing SQL condition to prevent selecting migrated children!). 

It looks there is already a check for that: https://github.com/FluidTYPO3/flux/blob/development/class.ext_update.php#L133-L144

I've tested with a minimal DB set and noticed that the migration wizard run through. But with my total data set I noticed the following: It seems the `dev-development` branch handles stuff a little different than the latest TER release and crashes when "possibly conflicting records" are still present. If I delete those (in my case already deleted `tt_content` records but about half of 25000 records in total) the migration wizard finishes without errors. Stupid me ignored this as "what could go wrong, those records are deleted anyway". "Ignore this check - at your own risk!" should be changed in "Without clean up the migration may fail!" ;-) 

Regarding the other error (`Argument 1 passed to Grid::buildExtendedBackendLayoutArray() must be of the type integer, array given, called in BackendLayoutView.php on line 76`): It seems this issue was caused by a custom `tt_content_drawItem` class using following hook: `$GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['cms/layout/class.tx_cms_layout.php']['tt_content_drawItem']`. The class is based upon what EXT:mask_export generates: https://github.com/IchHabRecht/mask_export/blob/master/Classes/Aggregate/BackendPreviewAggregate.php#L166 The marked line is what causes the error in flux.


So in conclusion: 
* In latest `dev-development` branch the SQL error is gone
* One REALLY should clean up his DB before running the migration
* Be careful with custom `tt_content_drawItem` classes

**Comment by NamelessCoder** [comment]
Thanks again for the awesome detective work here - makes it so much easier to target exactly the right places because the entire context is understood!

I've created two patches that I think may improve the situation. Since the described third-party usage of FormEngine expanded records is more common than we might like, I've chosen to implement a little edge-case support that handles that case even though it's not that common to see, since normally these two components - FormEngine and BackendLayoutView - are not directly related.

The second one switches the restriction removal from the query selecting records to migrate so it no longer removes the "deleted" restriction. After reading and considering, there's no really good reason to migrate those records - and they are definitely the most likely to cause problems with unresolvable files, bad parent/child relationships etc.

And since you've been contributing for so long and I trust your judgement, I've added you as collaborator which means that if you agree with those two pull requests you are welcome to merge them and get them in `development` without having to wait. It also gives you the option to push directly to development in emergencies but we should try to open PRs when the matter isn't urgent.

Thanks for sticking with Flux for so many years now!

**Comment by fnagel** [comment]
I've already tested and confirmed the patch for fixing the usage of custom FormEngines. Works as expected! In addition I did some testing with the changed migration DB queries. Looks almost good to me and helps as no manually clean up is needed (at least in my case). Please see the PRs for more info!

Thanks a lot for adding me to the list of collaborators and the the confidence you have in me. I'm flattered :-)  And *thank you* for all the years maintain Flux & Co!

**Comment by NamelessCoder** [comment]
Referenced PR is merged and I think released as well. Time to close this issue as "solved" :)
